examples/example-server-lib/sw_splash.cpp

```diff
@@ -18,92 +18,319 @@
+        void* data,
+        struct wl_registry* registry,
+        uint32_t id,
+        char const* interface,
+        uint32_t version);
+
+    static void global_remove(
+        void* data,
+        struct wl_registry* registry,
+        uint32_t name);
+};
 
-    for (unsigned int i = 0; i < num_formats; ++i)
+void Globals::new_global(
+    void* data,
+    struct wl_registry* registry,
+    uint32_t id,
+    char const* interface,
+    uint32_t version)
+{
+    (void)version;
+    Globals* self = static_cast<decltype(self)>(data);
+
+    if (strcmp(interface, "wl_compositor") == 0)
     {
-        MirPixelFormat cur_pf = pixel_formats[i];
-        if (cur_pf == mir_pixel_format_abgr_8888 ||
-            cur_pf == mir_pixel_format_argb_8888)
-        {
-            return cur_pf;
-        }
+        self->compositor =
+            static_cast<decltype(self->compositor)>(wl_registry_bind(registry, id, &wl_compositor_interface, 3));
     }
+    else if (strcmp(interface, "wl_shm") == 0)
+    {
+        self->shm = static_cast<decltype(self->shm)>(wl_registry_bind(registry, id, &wl_shm_interface, 1));
+        // Normally we'd add a listener to pick up the supported formats here
+        // As luck would have it, I know that argb8888 is the only format we support :)
+    }
+    else if (strcmp(interface, "wl_seat") == 0)
+    {
+        self->seat = static_cast<decltype(self->seat)>(wl_registry_bind(registry, id, &wl_seat_interface, 4));
+    }
+    else if (strcmp(interface, "wl_output") == 0)
+    {
+        self->output = static_cast<decltype(self->output)>(wl_registry_bind(registry, id, &wl_output_interface, 2));
+    }
+    else if (strcmp(interface, "wl_shell") == 0)
+    {
+        self->shell = static_cast<decltype(self->shell)>(wl_registry_bind(registry, id, &wl_shell_interface, 1));
+    }
+}
+
+void Globals::global_remove(
+    void* data,
+    struct wl_registry* registry,
+    uint32_t name)
+{
+    (void)data;
+    (void)registry;
+    (void)name;
+}
+
+void Globals::init(struct wl_display* display)
+{
+    wl_registry_listener const registry_listener = {
+        new_global,
+        global_remove
+    };
+
+    auto const registry = make_scoped(wl_display_get_registry(display), &wl_registry_destroy);
+
+    wl_registry_add_listener(registry.get(), &registry_listener, this);
+    wl_display_roundtrip(display);
+}
+
+struct OutputInfo
+{
+    int32_t x;
+    int32_t y;
+    int32_t width;
+    int32_t height;
+};
 
-    for (unsigned int i = 0; i < num_formats; ++i)
+using Outputs = std::map<struct wl_output*, OutputInfo>;
+
+void output_geometry(void *data,
+    struct wl_output *wl_output,
+    int32_t x,
+    int32_t y,
+    int32_t /*physical_width*/,
+    int32_t /*physical_height*/,
+    int32_t /*subpixel*/,
+    const char */*make*/,
+    const char */*model*/,
+    int32_t /*transform*/)
+{
+    Outputs* outputs = static_cast<decltype(outputs)>(data);
+
+    auto const& output = outputs->find(wl_output);
```

Comment:
I'm not a huge fan of taking references to temporaries.

Yes, I know it works, but it continues to be weird, particularly here where it serves no purpose.